### PR TITLE
Add JSON Renderer for Report

### LINKED ISSUE #404: Generate JSON report
Hi,

I'm using the latest version of PHPMD, but i can't use the command line to generate json report for violations.

Is it possible or not yet ?

Best regards.

### PR DESCRIPTION
**Pull request by exploitfate**
Close #404

### CODE REVIEW

**Comment by quentincurtet** [comment]
When do you think this will be merge with the actual version ?

File: src/main/php/PHPMD/Renderer/JsonRenderer.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace PHPMD\Renderer;
+
+use PHPMD\AbstractRenderer;
+use PHPMD\PHPMD;
+use PHPMD\Report;
+use PHPMD\RuleViolation;
+
+/**
+ * This class will render a json-report.
```
Comment (ravage84) [inline src/main/php/PHPMD/Renderer/JsonRenderer.php]: JSON report

File: src/main/php/PHPMD/Renderer/JsonRenderer.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * Temporary property that holds data for report
+     *
+     * @var array
+     */
+    private $jsonData;
+
+    /**
+     * @inheritdoc
```
Comment (ravage84) [inline src/main/php/PHPMD/Renderer/JsonRenderer.php]: ``{@inheritDoc}``
https://www.phpdoc.org/docs/latest/guides/inheritance.html

File: src/main/php/PHPMD/Renderer/JsonRenderer.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    public function renderReport(Report $report)
+    {
+        $writer = $this->getWriter();
+        $this->jsonData['version'] = PHPMD::VERSION;
```
Comment (ravage84) [inline src/main/php/PHPMD/Renderer/JsonRenderer.php]: These three lines could be extracted into a protected function, I guess.

File: src/main/php/PHPMD/Renderer/JsonRenderer.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                'description' => $violation->getDescription(),
+                'rule' => $rule->getName(),
+                'ruleSet' => $rule->getRuleSetName(),
+                'externalInfoUrl' => $rule->getExternalInfoUrl(),
+                'priority' => $rule->getPriority(),
+            );
+        }
+
+        $errorList = array();
+
```
Comment (ravage84) [inline src/main/php/PHPMD/Renderer/JsonRenderer.php]: Remove white line

File: src/main/php/PHPMD/Renderer/JsonRenderer.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        foreach ($report->getErrors() as $error) {
+            $errorList[] = array(
+                'fileName' => $error->getFile(),
+                'message' => $error->getMessage(),
+            );
+        }
+        $this->jsonData['files'] = array_values($filesList);
+        $this->jsonData['errors'] = $errorList;
+        $encodeOptions = JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_QUOT | JSON_HEX_AMP;
```
Comment (ravage84) [inline src/main/php/PHPMD/Renderer/JsonRenderer.php]: Please extract everthing that belongs to encoding the JSON into a protected method.

File: src/main/php/PHPMD/TextUI/CommandLineOptions.php [hunk 6]
```diff
@@ -387,7 +390,7 @@ protected function createXmlRenderer()
     }
 
     /**
-     * @return \PHPMD\Renderer\XMLRenderer
+     * @return \PHPMD\Renderer\TextRenderer
```
Comment (ravage84) [inline src/main/php/PHPMD/TextUI/CommandLineOptions.php]: :+1:

File: src/test/php/PHPMD/AbstractTest.php [hunk 7]
```diff
@@ -494,6 +494,39 @@ public static function assertXmlEquals($actualOutput, $expectedFileName)
     }
 
     /**
+     * Asserts the actual json output matches against the expected file.
```
Comment (ravage84) [inline src/test/php/PHPMD/AbstractTest.php]: JSON

File: src/test/php/PHPMD/Renderer/JsonRendererTest.php [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+ */
+
+namespace PHPMD\Renderer;
+
+use PHPMD\AbstractTest;
+use PHPMD\ProcessingError;
+use PHPMD\Stubs\WriterStub;
+
+/**
+ * Test case for the json renderer implementation.
```
Comment (ravage84) [inline src/test/php/PHPMD/Renderer/JsonRendererTest.php]: JSON

**Comment by ravage84** [comment]
I gues it needs some documenation additons, too.
But you can wait with them until we agreed on the implementation.

**Comment by ravage84** [comment]
@quentincurtet there is no plan as both maintainers of PHPMD are swamped in work/private stuff currently.

**Comment by exploitfate** [comment]
> I gues it needs some documenation additons, too.
> But you can wait with them until we agreed on the implementation.
OK
Should I create new PR for it?

**Comment by ravage84** [comment]
@exploitfate  sorry for not answering to your question. No you can append your changes to this PR, as it should be a shipable package in itself.

**Comment by ravage84** [comment]
Can you please rebase. Otherwise I cannot merge without merge conflicts

**Comment by exploitfate** [comment]
Hi,
I've added some documentations in README and fix command line options

**Comment by exploitfate** [comment]
Hi @ravage84 ,
I've done rebase. Can you check it?

**Comment by stormherz** [comment]
@ravage84 is there any plans on merging this one?

**Comment by kylekatarnls** [comment]
@MarkVaughn my changes dismissed your approval, note that I just add missing trailing new lines and fixed the conflict in README.rst so you can safely re-approve.

**Comment by ravage84** [comment]
I'll try to review it tonight.

File: src/main/php/PHPMD/Renderer/JSONRenderer.php [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+     * Encode report data to the JSON representation string
+     *
+     * @param $data array The report data
+     *
+     * @return string
+     */
+    private function encodeReport($data)
+    {
+        $encodeOptions = JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_QUOT | JSON_HEX_AMP;
+        // JSON_PRETTY_PRINT Available since PHP 5.4.0.
```
Comment (kylekatarnls) [inline src/main/php/PHPMD/Renderer/JSONRenderer.php]: As far as I know, we don't support PHP < 5.4:
https://github.com/phpmd/phpmd/blob/master/composer.json#L38

And indeed JSON_PRETTY_PRINT appears in 5.4.0 so we can directly add it to the options list.

File: src/test/php/PHPMD/AbstractTest.php [hunk 10]
```diff
@@ -480,6 +480,39 @@ public static function assertXmlEquals($actualOutput, $expectedFileName)
...
+
+        $expected = str_replace(
+            '#{rootDirectory}',
+            self::$filesDirectory,
+            file_get_contents(self::createFileUri($expectedFileName))
+        );
+
+        $expected = str_replace('_DS_', DIRECTORY_SEPARATOR, $expected);
+
+        self::assertJsonStringEqualsJsonString($expected, json_encode($actual));
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/AbstractTest.php]: As `assertJsonEquals` call `self::assertJsonStringEqualsJsonString` statically and do not use `$this` it should be called statically (and so declared as static function).

Else it should call `$this->assertJsonStringEqualsJsonString`. Seems we use `$this->assert...` everywhere else.

File: src/test/php/PHPMD/Renderer/JSONRendererTest.php [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+     * @return void
+     */
+    public function testRendererCreatesExpectedNumberOfJsonElements()
+    {
+        $writer = new WriterStub();
+        
+        $violations = array(
+            $this->getRuleViolationMock('/bar.php'),
+            $this->getRuleViolationMock('/foo.php'),
+            $this->getRuleViolationMock('/bar.php'), // TODO Set with description "foo <?php bar".
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/Renderer/JSONRendererTest.php]: @ravage84 I will add the descriptions in tests.

File: src/test/php/PHPMD/AbstractTest.php (hunk 10 above)
Comment (ravage84) [reply src/test/php/PHPMD/AbstractTest.php]: I made it static. We can address the whole static/dynamic mess some other time, some other place. As long as it works...

File: src/test/php/PHPMD/Renderer/JSONRendererTest.php (hunk 11 above)
Comment (ravage84) [reply src/test/php/PHPMD/Renderer/JSONRendererTest.php]: We can do that post-merge.

Comment (kylekatarnls) [reply src/test/php/PHPMD/Renderer/JSONRendererTest.php]: OK

**Comment by kylekatarnls** [comment]
@ravage84 Do we need to release a 2.6.2? If we merge 2.6.2 fixes in 2.7.0 and release 2.7.0 directly, we can merge this, else we should create a temporary branch or lock this PR so it's not merged in master.

So for now I approved but I keep my review as pending so it can not be merged for now in master.

**Comment by ravage84** [comment]
> @ravage84 Do we need to release a 2.6.2? If we merge 2.6.2 fixes in 2.7.0 and release 2.7.0 directly, we can merge this, else we should create a temporary branch or lock this PR so it's not merged in master.

No, we do not need. I just created that milestone to have more control over the issues. I wanted to take care of the issues reorganisation, next.

I could merge, funny enough.

**Pull request merged**